core/src/main/java/org/apache/iceberg/BaseFile.java

```diff
@@ -0,0 +1,389 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import com.google.common.base.MoreObjects;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import java.io.Serializable;
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import org.apache.avro.generic.IndexedRecord;
+import org.apache.avro.specific.SpecificData;
+import org.apache.iceberg.avro.AvroSchemaUtil;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.Types;
+import org.apache.iceberg.util.ByteBuffers;
+
+/**
+ * Base class for both {@link DataFile} and {@link DeleteFile}.
+ */
+abstract class BaseFile<F>
+    implements ContentFile<F>, IndexedRecord, StructLike, SpecificData.SchemaConstructable, Serializable {
+  static final Types.StructType EMPTY_STRUCT_TYPE = Types.StructType.of();
+  static final PartitionData EMPTY_PARTITION_DATA = new PartitionData(EMPTY_STRUCT_TYPE) {
+    @Override
+    public PartitionData copy() {
+      return this; // this does not change
+    }
+  };
+
+  private int[] fromProjectionPos;
+  private Types.StructType partitionType;
+
+  private FileContent content = FileContent.DATA;
```

Comment:
I see the comment says that the BaseFile is the base class for DataFile and DeleteFile,  is it suitable to make the FileContent use FileContent.DATA by default ?  Just curious.

Reply:
I am also curious whether we can rely on `content()` defined in `ContentFile` as both `DeleteFile` and `DataFile` override that.

Reply:
The FileContent for DeleteFile could be either `POSITION_DELETES` or `EQUALITY_DELETES` so we need to store it. DataFile always overrides it though.

Reply:
We actually need a field as `DeleteFile`s can be either positional or equality.